### PR TITLE
refactor(nyz): audio language model RL pipeline

### PR DESCRIPTION
**Pull request by PaParaZz1**
- [x] sglang pipeline
- [x] vllm pipeline
- [x] update doc
- [x] Qwen2-Audio/Qwen2.5-Omni adaption

### CODE REVIEW

File: examples/r1_aqa/data_preprocess/avqa.py [hunk 1]
```diff
@@ -82,28 +82,31 @@ def build_prompt_and_solution(
 
     # Build question template (matches R1-AQA)
     choice_str = f"Please choose the answer from the following options: {multi_choice}."
+    # There should be a space between <answer> and </answer>
```
Comment (puyuan1996) [inline examples/r1_aqa/data_preprocess/avqa.py:85]: 这个对结果影响很大吗？注释里面讲加不加空格对性能的影响是否更好呢

File: examples/r1_aqa/data_preprocess/avqa.py [hunk 2]
```diff
@@ -238,9 +241,7 @@ def preprocess_avqa(
 
 
 if __name__ == "__main__":
-    parser = argparse.ArgumentParser(
-        description="Preprocess AVQA dataset (R1-AQA format) for LightRFT training"
-    )
+    parser = argparse.ArgumentParser(description="Preprocess AVQA dataset (R1-AQA format) for LightRFT training")
```
Comment (puyuan1996) [inline examples/r1_aqa/data_preprocess/avqa.py:244]: 是否可以加下AVQA的链接呢

File: examples/r1_aqa/audio_dataset.py [hunk 3]
```diff
@@ -1,75 +1,47 @@
...
-    :param dataset: Underlying HuggingFace dataset.
-    :param tokenizer: HuggingFace tokenizer.
-    :param processor: Qwen2-Audio processor (for chat template + feature extraction).
-    :param max_length: Maximum prompt length.
-    :param strategy: LightRFT strategy object.
-    :param input_template: Optional template for formatting input text.
+    Each item returns ``(prompt_text, audio_payload, reference, label)`` where:
+    - ``prompt_text`` is rendered through the Qwen2-Audio chat template
+    - ``audio_payload`` is kept as raw waveform + sampling rate for rollout-side processing
+    - ``reference`` and ``label`` are passed through to reward computation
```
Comment (puyuan1996) [inline examples/r1_aqa/audio_dataset.py:49]: 为什么删掉:param dataset:这些呢

File: examples/r1_aqa/audio_dataset.py [hunk 4]
```diff
@@ -153,446 +112,40 @@ def __getitem__(self, idx: int) -> Tuple[str, Any, str, str]:
...
-        self,
-        all_prompts: List[str],
-        all_images: List,  # Actually audio data: List[Optional[Tuple[np.ndarray, int]]]
-        all_references: Optional[List[str]],
-        images_num: List[int],
-        n_samples_per_prompt: int,
-        all_videos: Optional[List] = None,
-        videos_num: Optional[List[int]] = None,
-    ) -> EasyDict:
+    @staticmethod
```
Comment (puyuan1996) [inline examples/r1_aqa/audio_dataset.py:174]: AudioMultimodalProcessor是没有用到了吗

File: examples/r1_aqa/eval.py [hunk 5]
```diff
@@ -58,10 +66,8 @@ def build_message(obj_dict: Dict, audio_dir: Optional[str] = None) -> list:
     :return: Chat messages list.
     """
     choice_str = f"Please choose the answer from the following options: {obj_dict['choices']}."
-    question_template = (
-        f"{obj_dict['question']} {choice_str} "
-        "Output the final answer in <answer></answer>."
-    )
+    question_template = (f"{obj_dict['question']} {choice_str} "
+                         "Output the final answer in <answer></answer>.")
```
Comment (puyuan1996) [inline examples/r1_aqa/eval.py]: 这里的 <answer></answer>中间没有空格？

File: examples/r1_aqa/reward_models_utils.py [hunk 6]
```diff
@@ -134,10 +136,26 @@ def avqa_combined_reward_fn(
     return total_r, acc_r, fmt_r
 
 
+def clean_solution(sol: str) -> str:
+    # <|im_start|>assistantat sea<|im_end|>
```
Comment (puyuan1996) [inline examples/r1_aqa/reward_models_utils.py]: 这行不需要吧？

File: lightrft/models/loss.py [hunk 7]
```diff
@@ -176,6 +176,64 @@ def __init__(
...
+        self,
+        *,
+        stats_mask: torch.Tensor,
+        advantages: torch.Tensor,
+        logprob_delta: torch.Tensor,
+        token_loss: torch.Tensor,
+        log_probs: Optional[torch.Tensor] = None,
+        old_log_probs: Optional[torch.Tensor] = None,
+        ratio: Optional[torch.Tensor] = None,
+    ) -> None:
```
Comment (puyuan1996) [inline lightrft/models/loss.py:238]: 增加下docstring?

File: lightrft/trainer/ppo_trainer_vl.py [hunk 8]
```diff
@@ -300,106 +414,73 @@ def fit(
...
-                        all_labels=rand_labels,
-                        **self.generate_kwargs
-                    )
-                ):
-                    if i == 0:
-                        output = self.tokenizer.batch_decode(
-                            experience.sequences[0].unsqueeze(0), skip_special_tokens=True
-                        )
-                        self.strategy.print("collect phase: experience.sequences w skip_special_tokens: ", output)
-                        self.strategy.print(
```
Comment (puyuan1996) [inline lightrft/trainer/ppo_trainer_vl.py:389]: 启动脚本中加个debug的option 如果打开就print这里的信息吧？方便debug分析

File: lightrft/trainer/ppo_trainer_vl.py [hunk 9]
```diff
@@ -622,82 +682,124 @@ def training_step(self,
...
+            attention_mask = torch.cat(
+                [torch.full_like(seq, idx + 1) for idx, seq in enumerate(experience.sequences)],
+                dim=0,
+            ).unsqueeze(0)
             if self.args.use_kl_loss and experience.base_action_log_probs is not None:
                 base_action_log_probs = torch.cat(experience.base_action_log_probs, dim=0).unsqueeze(0)
         else:
             sequences = experience.sequences
-
-            pixel_values = experience.pixel_values
```
Comment (puyuan1996) [inline lightrft/trainer/ppo_trainer_vl.py:696]: 为什么 pixel_values相关都删除了呀

File: lightrft/trainer/ppo_trainer_vl.py [hunk 10]
```diff
@@ -763,27 +844,18 @@ def training_step_actor(self,
...
                 kl = compute_approx_kl(
                     action_log_probs,
                     base_action_log_probs,
                     experience.action_mask,
                     kl_estimator=self.args.kl_estimator,
                 )
-
-                # [Protection measure 2] Per-token KL Clamping
-                # NOTE: Adding this causes svkng training to not converge
-                # kl = torch.clamp(kl, min=0.0, max=20.0)
```
Comment (puyuan1996) [inline lightrft/trainer/ppo_trainer_vl.py:777]: 这个保留吧？之前发现的关键信息

File: lightrft/trainer/ppo_trainer_vl.py [hunk 11]
```diff
@@ -966,162 +988,127 @@ def ensure_device_and_contiguous(tensor, name="tensor"):
...
                     eval_logs["eval/global_step"] = self.eval_step_counter
-                    # Reference to main training step
                     eval_logs["eval/train_step"] = global_step
                     eval_logs["eval/episode"] = episode
-
-                    # IMPORTANT:
-                    # Use wandb_log_counter to ensure eval has a unique system step
-                    # This prevents eval metrics from being overwritten by train metrics
-                    # The plots will still use eval/global_step as X-axis due to define_metric
                     self.wandb_log_counter += 1
```
Comment (puyuan1996) [inline lightrft/trainer/ppo_trainer_vl.py:1326]: 关键的注释保留吧？

**Review by puyuan1996 — CHANGES_REQUESTED** [review]

File: examples/r1_aqa/README.md [hunk 12]
```diff
@@ -155,18 +197,14 @@ The reward function automatically handles both modes. When `enable_think=True`,
...
-- `image_grid_thw` → `feature_attention_mask`
-- `raw_images` → raw audio tuples `(np.array, sr)`
-- `multi_modal_data["image"]` → `multi_modal_data["audio"]`
-
-This is done via targeted monkey patches in `audio_dataset.py` rather than modifying core LightRFT code.
+### 2. Native Audio Rollout Path
+Audio RL now uses a dedicated rollout path in core LightRFT code:
+- raw audio payloads stay on the generation side and are passed to SGLang as `audio_data`
+- processed mel features are stored explicitly as `audio_values`
+- Qwen2-Audio feature masking is stored explicitly as `feature_attention_mask`
```
Comment (puyuan1996) [inline examples/r1_aqa/README.md:207]: 是否可以增加中文readme

File: examples/r1_aqa/data_preprocess/avqa.py (hunk 1 above)
Comment (PaParaZz1) [reply examples/r1_aqa/data_preprocess/avqa.py:85]: 对 qwen2-audio 的初始收敛速度影响很大，`should be` 就是说明要加这个空格的意思

File: examples/r1_aqa/data_preprocess/avqa.py (hunk 2 above)
Comment (PaParaZz1) [reply examples/r1_aqa/data_preprocess/avqa.py:244]: 这个文件一开头的注释有 AVQA 的链接

File: examples/r1_aqa/audio_dataset.py (hunk 3 above)
Comment (PaParaZz1) [reply examples/r1_aqa/audio_dataset.py:49]: 补上了

File: examples/r1_aqa/audio_dataset.py (hunk 4 above)
Comment (PaParaZz1) [reply examples/r1_aqa/audio_dataset.py:174]: 对，重构删掉了

File: examples/r1_aqa/eval.py (hunk 5 above)
Comment (PaParaZz1) [reply examples/r1_aqa/eval.py]: 改了

File: examples/r1_aqa/reward_models_utils.py (hunk 6 above)
Comment (PaParaZz1) [reply examples/r1_aqa/reward_models_utils.py]: 删掉了

File: lightrft/models/loss.py (hunk 7 above)
Comment (PaParaZz1) [reply lightrft/models/loss.py:238]: 添加了

File: examples/r1_aqa/README.md (hunk 12 above)
Comment (PaParaZz1) [reply examples/r1_aqa/README.md:207]: 加上了

File: lightrft/trainer/ppo_trainer_vl.py (hunk 11 above)
Comment (PaParaZz1) [reply lightrft/trainer/ppo_trainer_vl.py:1326]: 加回去了

File: lightrft/trainer/ppo_trainer_vl.py (hunk 8 above)
Comment (PaParaZz1) [reply lightrft/trainer/ppo_trainer_vl.py:389]: 这不是这个 PR 的功能，应该在其他 polish PR 弄

File: lightrft/trainer/ppo_trainer_vl.py (hunk 10 above)
Comment (PaParaZz1) [reply lightrft/trainer/ppo_trainer_vl.py:777]: 加回去了

File: lightrft/trainer/ppo_trainer_vl.py (hunk 9 above)
Comment (PaParaZz1) [reply lightrft/trainer/ppo_trainer_vl.py:696]: 是放在 `_build_model_kwargs` 方法里统一实现了，这样对各种模态都能处理
